### PR TITLE
[DLS-6528] - WIP

### PR DESCRIPTION
**Pull request by JakeReid2020**

### CODE REVIEW

File: app/models/AddASmallProducer.scala [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package models
+
+import play.api.libs.json._
+
+case class AddASmallProducer (producerName:Option[String],referenceNumber:String,lowBand: Long, highBand: Long)
```
Comment (priyankadhiman3006) [inline app/models/AddASmallProducer.scala:21]: Shouldn't this be AddSmallProducer?

Comment (JakeReid2020) [reply app/models/AddASmallProducer.scala:21]: I used the title of the ticket but can change it

**Comment by priyankadhiman3006** [comment]
image
The error message is appearing twice 
2) When an invalid SDIL NUMBER format is entered , there is internal server error and user is redirected to sorry we'r experiencing technical difficulties

File: app/forms/AddASmallProducerFormProvider.scala [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                                referenceNumberList: List[String], errorKey3: String,
+                                returnPeriod:ReturnPeriod,errorKey4: String): Constraint[String] =
+       Constraint {
+         case str if !str.matches(regex) =>
+           Invalid(errorKey1, regex)
+         case str if str.matches(referenceNumber) =>
+           Invalid(errorKey2, referenceNumber)
+         case str if referenceNumberList.contains(str) =>
+           Invalid(errorKey3, referenceNumberList)
+         case str if (Await.result(checkSmallProducerStatus(str,returnPeriod),20.seconds).getOrElse(true)).equals(false)  =>
```
Comment (priyankadhiman3006) [inline app/forms/AddASmallProducerFormProvider.scala:60]: why are we using the await in here. it's not recommended to use in actual code other than unit testing

File: generated-test/forms/smallProducerDetailsFormProviderSpec.scala [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+class smallProducerDetailsFormProviderSpec extends BooleanFieldBehaviours {
+
+  val requiredKey = "smallProducerDetails.error.required"
+  val invalidKey = "error.boolean"
+
+  val form = new smallProducerDetailsFormProvider()()
+
+  ".value" - {
+
+    val fieldName = "value"
```
Comment (priyankadhiman3006) [inline generated-test/forms/smallProducerDetailsFormProviderSpec.scala:15]: you should add more tests here like the constraints  we have on the SDIl text box

File: test/forms/AddASmallProducerFormProviderSpec.scala [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    val fieldName = "referenceNumber"
+    val requiredKey = "addASmallProducer.error.referenceNumber.required"
+
+    behave like mandatoryField(
+      form,
+      fieldName,
+      requiredError = FormError(fieldName, requiredKey)
+    )
+
+    behave like mandatoryField(
```
Comment (priyankadhiman3006) [inline test/forms/AddASmallProducerFormProviderSpec.scala:94]: duplicate of the above test

File: test/forms/AddASmallProducerFormProviderSpec.scala [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+      requiredError = FormError(fieldName, invalid)
+    )
+
+    behave like sameRefNumber(
+      form,
+      fieldName,
+      requiredError = FormError(fieldName, same)
+    )
+
+    behave like ExistingRefNumber(
```
Comment (priyankadhiman3006) [inline test/forms/AddASmallProducerFormProviderSpec.scala]: `ExistingRefNumber` make e small case

File: test/forms/behaviours/FieldBehaviours.scala [hunk 6]
```diff
@@ -55,4 +55,89 @@ trait FieldBehaviours extends FormSpec with ScalaCheckPropertyChecks with Genera
...
+    }
+
+    "not bind SameRef values" in {
+
+      val result = form.bind(Map(fieldName -> "XKSDIL000000022")).apply(fieldName)
+      result.errors mustEqual Seq(requiredError)
+    }
+  }
+
+  def ExistingRefNumber(form: Form[_],
```
Comment (priyankadhiman3006) [inline test/forms/behaviours/FieldBehaviours.scala]: e should be lower case

File: it/controllers/AddASmallProducerControllerIntergrationSpec.scala [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package controllers
+
+import controllers.testSupport.{Specifications, TestConfiguration}
+import play.api.libs.ws.DefaultWSCookie
+import play.api.test.WsTestClient
+
+class AddASmallProducerControllerIntergrationSpec extends Specifications with TestConfiguration {
+  "AddASmallProducerController" should {
+    "Ask for if user wants to register any UK warehouses where user used to store liable drinks" in {
```
Comment (priyankadhiman3006) [inline it/controllers/AddASmallProducerControllerIntergrationSpec.scala]: Is this correct? Shouldn't it be is user wants to add a small producer?

**Review by priyankadhiman3006 — APPROVED** [review]

**Pull request merged**